storage/storage.c

```diff
@@ -714,24 +818,65 @@ static secbool __wur derive_kek_unlock(
     const uint8_t *pin, size_t pin_len, const uint8_t *storage_salt,
     const uint8_t *ext_salt, uint8_t stretched_pin[SHA256_DIGEST_LENGTH],
     secbool privileged_bhk) {
-  stretch_pin(pin, pin_len, storage_salt, ext_salt, stretched_pin,
-              privileged_bhk);
+  secbool ret = secfalse;
+  mcu_stretch_pin(pin, pin_len, storage_salt, ext_salt, stretched_pin,
+                  privileged_bhk);
+#if USE_OPTIGA || USE_TROPIC
+  uint32_t pin_index = 0;
+#if STRETCHED_PIN_COUNT > 1
+  uint32_t pin_fails = 0;
+  if (sectrue != pin_get_fails(&pin_fails)) {
+    goto cleanup;
+  }
+  pin_index = pin_fails - 1;  // The counter has already been incremented
+#endif
+#endif
 #if USE_OPTIGA
-  optiga_pin_result ret = optiga_pin_verify(ui_progress, stretched_pin);
-  if (ret != OPTIGA_PIN_SUCCESS) {
+  optiga_pin_result optiga_ret = optiga_stretch_pin(ui_progress, stretched_pin);
+  if (optiga_ret != OPTIGA_PIN_SUCCESS) {
     memzero(stretched_pin, SHA256_DIGEST_LENGTH);
 
-    if (ret == OPTIGA_PIN_COUNTER_EXCEEDED) {
+    if (optiga_ret == OPTIGA_PIN_COUNTER_EXCEEDED) {
       // Unreachable code. Wipe should have already been triggered in unlock().
       storage_wipe();
       show_pin_too_many_screen();
     }
-    ensure(ret == OPTIGA_PIN_INVALID ? sectrue : secfalse,
+    ensure(optiga_ret == OPTIGA_PIN_INVALID ? sectrue : secfalse,
```

Comment:
`optiga_stretch_pin()` can't return `OPTIGA_PIN_INVALID`, can it? So does that mean we always RSOD?
- If yes, then let's be more explicit about it.
- Should we RSOD? I realize we do below with `optiga_pin_verify()`, so that's consistent, but we don't RSOD for any of the other functions when they fail e.g. `tropic_stretch_pin()`, so that doesn't seem consistent. I don't know what the right answer is, but we should make the behavior consistent unless there is a reason not to.